metadata-ingestion/src/datahub/ingestion/source/sql/bigquery.py

```diff
@@ -140,20 +144,59 @@ def get_view_definition(self, connection, view_name, schema=None, **kw):
 assert pybigquery.sqlalchemy_bigquery._type_map
 
 
+class BigQueryCredential(ConfigModel):
+    project_id: str
+    private_key_id: str
+    private_key: str
+    client_email: str
+    client_id: str
+    auth_uri: str = "https://accounts.google.com/o/oauth2/auth"
+    token_uri: str = "https://oauth2.googleapis.com/token"
+    auth_provider_x509_cert_url: str = "https://www.googleapis.com/oauth2/v1/certs"
+    type: str = "service_account"
+    client_x509_cert_url: Optional[str]
+
+    def __init__(self, **data: Any):
+        super().__init__(**data)  # type: ignore
+        if not self.client_x509_cert_url:
+            self.client_x509_cert_url = (
+                f"https://www.googleapis.com/robot/v1/metadata/x509/{self.client_email}"
+            )
+
+
+def create_credential_temp_file(credential: BigQueryCredential) -> str:
+    with tempfile.NamedTemporaryFile(delete=False) as fp:
+        cred_json = json.dumps(credential.dict(), indent=4, separators=(",", ": "))
+        fp.write(cred_json.encode())
+        return fp.name
+
+
 class BigQueryConfig(BaseTimeWindowConfig, SQLAlchemyConfig):
     scheme: str = "bigquery"
     project_id: Optional[str] = None
 
     log_page_size: Optional[pydantic.PositiveInt] = 1000
+    credential: Optional[BigQueryCredential]
     # extra_client_options, include_table_lineage and max_query_duration are relevant only when computing the lineage.
     extra_client_options: Dict[str, Any] = {}
     include_table_lineage: Optional[bool] = True
     max_query_duration: timedelta = timedelta(minutes=15)
 
+    credentials_path: Optional[str] = None
     bigquery_audit_metadata_datasets: Optional[List[str]] = None
     use_exported_bigquery_audit_metadata: bool = False
     use_date_sharded_audit_log_tables: bool = False
 
+    def __init__(self, **data: Any):
+        super().__init__(**data)
+
+        if self.credential:
+            self.credentials_path = create_credential_temp_file(self.credential)
+            logger.debug(
+                f"Creating temporary credential file at {self.credentials_path}"
+            )
+            os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = self.credentials_path
```

Comment:
Is this strictly required? It is strange we need to set an env variable

Reply:
We are using various APIs (sqlalchemy, BigQueryClient (https://cloud.google.com/bigquery/docs/reference/libraries#setting_up_authentication) and GCPLoggingClient (https://cloud.google.com/logging/docs/reference/libraries#setting_up_authentication), and with this way, we can authenticate with all of these.
I think it is simple and safe as this environment variable only lives during the life of the python app but works for all the APIs.